### ISSUE TITLE
Custom type in subquery

### ISSUE THREAD

**Issue by sagarrohat** [line/kotlin-jdsl#394]
I have a custom type `ACustomType` that extends `AbstractSingleColumnStandardBasicType` and I would like to use it like this:

```
val customTypeSubQuery = queryFactory.subquery<ACustomType> {
    select(col(AnEntity::aColumnOfCustomType))
    from(entity(AnEntity::class))
    where(col(AnEntity::anotherColumn).equal(someValue))
}

queryFactory.deleteQuery<AnotherEntity> {
    where(col(AnotherEntity::aColumnOfCustomType).`in`(customTypeSubQuery ))
}.executeUpdate()
```

The above code does not delete rows from the AnotherEntity table but if I fetch by using selectQuery then it works. I think it has something to do with Type Conversion

**Comment by shouwn** [comment]
@cj848 Could you check this? 
I'm not sure how the Delete query is implemented.

**Comment by shouwn** [comment]
Hi @sagarrohat. Sorry for late reply. 😢 

I executed a delete query with subquery using following entities.
```kotlin
@Entity
@Table(name = "an_entity")
class AnEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", length = 20, updatable = false, nullable = false)
    val id: Long,

    @Convert(converter = ACustomTypeConverter::class)
    val aColumnOfCustomType: ACustomType,
) {
    override fun equals(other: Any?) = JpaEntityUtils.equalsWith(this, other) { it.id }
    override fun hashCode() = Objects.hash(this.id)

    override fun toString(): String = "AnEntity(id=$id)"
}

@Entity
@Table(name = "another_entity")
class AnotherEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", length = 20, updatable = false, nullable = false)
    val id: Long,

    @Convert(converter = ACustomTypeConverter::class)
    val aColumnOfCustomType: ACustomType,
) {
    override fun equals(other: Any?) = JpaEntityUtils.equalsWith(this, other) { it.id }
    override fun hashCode() = Objects.hash(this.id)

    override fun toString(): String = "AnotherEntity(id=$id)"
}

abstract class AbstractSingleColumnStandardBasicType {
    abstract val name: String
}

data class ACustomType(
    override val name: String,
    val value: String,
) : AbstractSingleColumnStandardBasicType()

@Converter
class ACustomTypeConverter : AttributeConverter<ACustomType?, String?> {
    override fun convertToDatabaseColumn(attribute: ACustomType?): String? {
        return attribute?.let { "${it.name}:${it.value}" }
    }

    override fun convertToEntityAttribute(dbData: String?): ACustomType? {
        return dbData?.let {
            val (name, value) = it.split(":")

            ACustomType(name, value)
        }
    }
}
```

```kotlin
private val name1 = "name1"
private val name2 = "name2"
private val name3 = "name3"

private val value1 = "value1"
private val value2 = "value2"
private val value3 = "value3"

@Test
fun delete() {
    val aCustomType1 = ACustomType(name1, value1)
    val aCustomType2 = ACustomType(name2, value2)
    val aCustomType3 = ACustomType(name3, value3)

    val anEntity1 = AnEntity(id = 0, aColumnOfCustomType = aCustomType1)
    val anEntity2 = AnEntity(id = 0, aColumnOfCustomType = aCustomType2)
    val anEntity3 = AnEntity(id = 0, aColumnOfCustomType = aCustomType3)

    val anotherEntity1 = AnotherEntity(id = 0, aColumnOfCustomType = aCustomType1)
    val anotherEntity2 = AnotherEntity(id = 0, aColumnOfCustomType = aCustomType2)
    val anotherEntity3 = AnotherEntity(id = 0, aColumnOfCustomType = aCustomType3)

    entityManager.persistAll(
        anEntity1, anEntity2, anEntity3,
        anotherEntity1, anotherEntity2, anotherEntity3,
    )

    val customTypeSubquery = queryFactory.subquery<ACustomType> {
        select(col(AnEntity::aColumnOfCustomType))
        from(entity(AnEntity::class))
        where(col(AnEntity::id).equal(anEntity1.id))
    }

    queryFactory.deleteQuery<AnotherEntity> {
        where(col(AnotherEntity::aColumnOfCustomType).`in`(customTypeSubquery))
    }.executeUpdate()

    val result = queryFactory.selectQuery<AnotherEntity> {
        select(AnotherEntity::class.java)
        from(AnotherEntity::class)
    }

    println(result.resultList)
}
```

```sql
delete from another_entity where a_column_of_custom_type in (select anentity1_.a_column_of_custom_type from an_entity anentity1_ where anentity1_.id=1)
```

And I checked that the entity I wanted was deleted. Maybe this was caused by the difference between my entity and yours. 

Could you explain your entity structure in more detail?

**Comment by anshuhishek** [comment]
//Hi @anshuhishek , Sorry for late reply. 😢

//I executed a query


import java.util.Objects
import javax.persistence.*

@Entity
@Table(name = "an_entity")
class AnEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", length = 20, updatable = false, nullable = false)
    val id: Long,

    @Convert(converter = ACustomTypeConverter::class)
    val aColumnOfCustomType: ACustomType,
) {
    override fun equals(other: Any?) = JpaEntityUtils.equalsWith(this, other) { it.id }
    override fun hashCode() = Objects.hash(this.id)

    override fun toString(): String = "AnEntity(id=$id)"
}

@Entity
@Table(name = "another_entity")
class AnotherEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", length = 20, updatable = false, nullable = false)
    val id: Long,

    @Convert(converter = ACustomTypeConverter::class)
    val aColumnOfCustomType: ACustomType,
) {
    override fun equals(other: Any?) = JpaEntityUtils.equalsWith(this, other) { it.id }
    override fun hashCode() = Objects.hash(this.id)

    override fun toString(): String = "AnotherEntity(id=$id)"
}

@Converter
class ACustomTypeConverter : AttributeConverter<ACustomType?, String?> {
    override fun convertToDatabaseColumn(attribute: ACustomType?): String? {
        return attribute?.let { "${it.name}:${it.value}" }
    }

    override fun convertToEntityAttribute(dbData: String?): ACustomType? {
        return dbData?.let {
            val (name, value) = it.split(":")

            ACustomType(name, value)
        }
    }
}

**Comment by shouwn** [comment]
Hi @anshuhishek 

I thought `AbstractSingleColumnStandardBasicType` was a class you created, but I see it's a class in hibernate. Could you explain in more detail how you implemented `ACustomType`? Because I've never used `AbstractSingleColumnStandardBasicType` before and I don't know how to pass parameters to the parent constructor.

And maybe the code you wrote is a copy of my code.

**Comment by shouwn** [comment]
I checked the link. 

I found that instead of using the AbstractSingleColumnStandardBasicType provided by Hibernate, you used your own implementation of AbstractSingleColumnStandardBasicType to implement the abstract type. 

Then... I'm still not sure why it fails, because I implemented an Entity with the same structure and it worked when I tested it.

Does not deleting a row mean that Hibernate or Eclipselink threw an error, or does it mean that the query was sent to the DB and the row was not deleted as a result of the DB query? If an error occurred, could you give me a stack trace, if a DB query was sent, could you give me the DB query sent via logs? 

If you're using Spring Boot, you can also try the Kotlin JDSL 3.0.0, which is still in development, and see if you can get it to work with JPQL.

[EmployeeRepository](https://github.com/line/kotlin-jdsl/blob/develop/3.0.0/example/spring-data-jpa/src/main/kotlin/com/linecorp/kotlinjdsl/example/jpql/spring/jpa/repository/employee/EmployeeRepository.kt#L7
)
[DeleteExample](https://github.com/line/kotlin-jdsl/blob/develop/3.0.0/example/spring-data-jpa/src/test/kotlin/com/linecorp/kotlinjdsl/example/jpql/spring/jpa/delete/DeleteExample.kt#L52)

You can use Kotlin JDSL 3.0.0 by adding dependencies and extending `KotlinJdslJpqlExecutor` to your `JpaRepository`. 

```kotlin
repositories {
    mavenCentral()
    maven(url = "https://oss.sonatype.org/content/repositories/snapshots")
}

dependencies {
    implementation("com.linecorp.kotlin-jdsl:jpql-dsl:3.0.0-SNAPSHOT")
    implementation("com.linecorp.kotlin-jdsl:jpql-render:3.0.0-SNAPSHOT")
    implementation("com.linecorp.kotlin-jdsl:spring-data-jpa-support:3.0.0-SNAPSHOT") // for Spring Boot 3.0.3 (or later)
    implementation("com.linecorp.kotlin-jdsl:spring-data-jpa-javax-support:3.0.0-SNAPSHOT") // for Spring Boot 2.7.10 (or later)
}
```

**Comment by anshuhishek** [comment]
ok i will try, 

actually, the absence of an error being thrown does not necessarily mean that the row was deleted. It means that the query was successfully executed and there were no exceptional issues with the execution of the query itself (such as syntax errors or other database-related errors). However, the query being executed successfully does not guarantee that a row was actually deleted from the database.

**Comment by shouwn** [comment]
I am closing this issue for now. If the issue is not resolved in 3.0.0, please reopen it. 

If you want to know the JPQL generated by Kotlin JDSL 3.0.0, you can see it in the debug log. 

Please check that the JPQL generated by Kotlin JDSL 3.0.0 looks like what you want, and that the actual query that is sent looks like what you want.
